core/network/hostport.go

```diff
@@ -13,208 +13,361 @@ import (
+		}
+	}
+
+	if len(selectedHostPorts) > 0 {
+		return selectedHostPorts, true
+	}
+
+	logger.Errorf("no hostPorts found in spaces %s", spaceInfos)
+	return hps, false
 }
 
-// SortHostPorts sorts the given HostPort slice according to the sortOrder of
-// each HostPort's embedded Address. See Address.sortOrder() for more info.
-func SortHostPorts(hps []HostPort) {
-	sort.Sort(hostPortsPreferringIPv4Slice(hps))
+// SelectInternal picks the best matching HostPorts that can be used as an
+// endpoint for juju internal communication and returns them in NetAddr form.
+// If there are no suitable addresses, an empty slice is returned.
+func (hps SpaceHostPorts) SelectInternal(machineLocal bool) []string {
+	indexes := bestAddressIndexes(len(hps), func(i int) Address {
+		return hps[i].SpaceAddress
+	}, internalAddressMatcher(machineLocal))
+
+	out := make([]string, 0, len(indexes))
+	for _, index := range indexes {
+		out = append(out, DialAddress(hps[index]))
+	}
+	return out
 }
 
-// FilterUnusableHostPorts returns a copy of the given HostPorts after
-// removing any addresses unlikely to be usable (ScopeMachineLocal or
-// ScopeLinkLocal).
-func FilterUnusableHostPorts(hps []HostPort) []HostPort {
-	filtered := make([]HostPort, 0, len(hps))
-	for _, hp := range hps {
-		switch hp.Scope {
-		case ScopeMachineLocal, ScopeLinkLocal:
-			continue
+// Less reports whether hp is ordered before hp2
+// according to the criteria used by SortHostPorts.
+func (hp SpaceHostPort) Less(hp2 SpaceHostPort) bool {
+	order1 := hp.sortOrder()
+	order2 := hp2.sortOrder()
+	if order1 == order2 {
+		if hp.SpaceAddress.Value == hp2.SpaceAddress.Value {
+			return hp.Port() < hp2.Port()
 		}
-		filtered = append(filtered, hp)
+		return hp.SpaceAddress.Value < hp2.SpaceAddress.Value
 	}
-	return filtered
+	return order1 < order2
 }
 
-// HostPortsToStrings converts each HostPort to string calling its
-// NetAddr() method.
-func HostPortsToStrings(hps []HostPort) []string {
-	result := make([]string, len(hps))
-	for i, hp := range hps {
-		result[i] = hp.NetAddr()
+// SpaceAddressesWithPort returns the input SpaceAddresses
+// all associated with the given port.
+func SpaceAddressesWithPort(addrs []SpaceAddress, port int) SpaceHostPorts {
+	hps := make([]SpaceHostPort, len(addrs))
+	for i, addr := range addrs {
+		hps[i] = SpaceHostPort{
+			SpaceAddress: addr,
+			NetPort:      NetPort(port),
+		}
 	}
-	return result
+	return hps
+}
+
+// NewSpaceHostPorts creates a list of SpaceHostPorts
+// from each input string address and port.
+func NewSpaceHostPorts(port int, addresses ...string) SpaceHostPorts {
+	hps := make([]SpaceHostPort, len(addresses))
+	for i, addr := range addresses {
+		hps[i] = SpaceHostPort{
+			SpaceAddress: NewSpaceAddress(addr),
+			NetPort:      NetPort(port),
+		}
+	}
+	return hps
+}
+
+type hostPortsPreferringIPv4Slice []SpaceHostPort
+
+func (hp hostPortsPreferringIPv4Slice) Len() int      { return len(hp) }
```

Comment:
Sorting could probably be done on `SpaceHostPorts` instead of this new type, but I have left it as is for the time-being.